### PR TITLE
[KED-2996] Feature/toggle bookmark behavior

### PR DESCRIPTION
**Pull request by tynandebold**
## Description

Resolves [KED-2996](https://jira.quantumblack.com/browse/KED-2996).

## Development notes

This PR enables all the bookmark functionality. There are three places where you can bookmark a run from:

- The runs list card
- The pipeline toolbar (when not in comparison view)
- From the kebab menu (when in comparison view)

Each area is wired up to now be able to toggle bookmarks.

I also wired up the close icon next to the kebab menu. That was a quick win and easy to do.

## QA notes

I chose not to include the bookmark help text on hover in the runs list. Here's a design screenshot:

image

I did this for two reasons: 

- By default, the help text gets cut off and can't extend outside of the sidebar. There's a lot of overhead here that would have made this seemingly simple change time consuming and, in my opinion, not worthwhile for our MVP approach.
- When I did have it there, though not working perfectly, it felt very intrusive with the tooltip coming up over and over and over again.

Separately, I also ordered the icons like this:

image

And not like this:

image

Isn't the close icon always on an end? I didn't feel right when it was in the middle. 

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: src/components/experiment-tracking/runs-list/runs-list.scss
Comment (rashidakanchwala) [inline src/components/experiment-tracking/runs-list/runs-list.scss:7]: nit - maybe we can remove the comments?

**Review by rashidakanchwala — APPROVED** [review]
LGTM! Thanks Tynan!!!

File: src/components/experiment-tracking/runs-list/runs-list.scss
Comment (tynandebold) [reply src/components/experiment-tracking/runs-list/runs-list.scss]: Of course. This is leftover. Great catch. Removed!

File: src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js
Comment (studioswong) [inline src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js:31]: This function is used in 2 places ( both here and `runMetadata.js` ) - I would extract that out to the main experiment wrapper to pass it down to both components instead.

File: src/components/experiment-tracking/run-metadata/run-metadata.js
Comment (studioswong) [inline src/components/experiment-tracking/run-metadata/run-metadata.js:31]: see above comment

File: src/components/experiment-tracking/run-metadata/run-metadata.test.js
Comment (studioswong) [inline src/components/experiment-tracking/run-metadata/run-metadata.test.js:114]: This test can be put back in once the  `useMutation` hook is moved up to the wrapper level.

On another note, please add relevant tests to cover the introduction of the new kebab menu and the bookmark icon.

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (studioswong) [inline src/components/experiment-tracking/runs-list-card/runs-list-card.js:29]: please rename this to something more descrption such as `onToggleBookmarkButton`, etc

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (studioswong) [inline src/components/experiment-tracking/runs-list-card/runs-list-card.js:87]: please add a test for testing the expected display of the different icons

**Review by studioswong — CHANGES_REQUESTED** [review]
Thanks for setting this up, the bookmark journey works well in general.

One thing I noticed.  - on toggling 'show changes' off the entire burger menu, including the toggle bookmark buttons, etc, also gets toggled off, while instead they should still exist and only the 'pin run' button should be toggled off. 

I'll continue to go through some of the code and leave further single comments if I see anything.

File: src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js
Comment (tynandebold) [reply src/components/experiment-tracking/experiment-primary-toolbar/experiment-primary-toolbar.js:31]: Good idea. Though instead of passing it as a prop I'm going to make it a custom hook so we can pass it around freely as we wish.

File: src/components/experiment-tracking/run-metadata/run-metadata.test.js
Comment (tynandebold) [reply src/components/experiment-tracking/run-metadata/run-metadata.test.js:114]: I wish this was true. The test still fails even if the mutation hook is moved. The hook itself still fires in this file, so it doesn't make a difference.

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (tynandebold) [reply src/components/experiment-tracking/runs-list-card/runs-list-card.js:29]: This function can either toggle the bookmark or fire the `onRunSelection` function, so it isn't strictly about the bookmark. 

I can change it to `onRunsListCardClick`. Is that ok?

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (tynandebold) [reply src/components/experiment-tracking/runs-list-card/runs-list-card.js:87]: Added.

**Comment by tynandebold** [comment]
> Thanks for setting this up, the bookmark journey works well in general.
> 
> One thing I noticed. - on toggling 'show changes' off the entire burger menu, including the toggle bookmark buttons, etc, also gets toggled off, while instead they should still exist and only the 'pin run' button should be toggled off.
> 
> I'll continue to go through some of the code and leave further single comments if I see anything.

Updated everything you asked for. Lmk if there's anything else :)

File: src/components/experiment-tracking/run-metadata/run-metadata.test.js
Comment (studioswong) [reply src/components/experiment-tracking/run-metadata/run-metadata.test.js:114]: I think this problem with `useMutation` would be something we need to resolve within this PR as this seems to restrict our ability to write tests, which I highly doubt it would be designed this way given how many users there are within the react community. 

Throwing a thought here - looking at the error it looks like the test was not able to locate the client, which you might need to wrap the wrapper with an ApolloProvider in the test?

Comment (studioswong) [reply src/components/experiment-tracking/run-metadata/run-metadata.test.js:114]: Also, it also does seem weird that this problem surfaces in this PR and not in the previous one when we introduced the mutation hook for the first time - I wonder what is the reason behind that?

it might either suggest that there might be a different setup, or perhaps there wasn't the right tests set up to properly test that 😆

File: src/apollo/mutations.js
Comment (tynandebold) [reply src/apollo/mutations.js]: No it wasn't that. It's just we were't actually mocking the `useUpdateRunDetails` mutation.

File: src/apollo/queries.js
Comment (limdauto) [inline src/apollo/queries.js:22]: not in scope of your PR but this query has duplicated `id` field. Can we quickly remove it here?

**Review by limdauto — APPROVED** [review]
I haven't tested the functionality but the code LGTM!

File: src/apollo/queries.js
Comment (tynandebold) [reply src/apollo/queries.js:21]: Absolutely. It's gone.

File: src/apollo/mutations.js
Comment (studioswong) [reply src/apollo/mutations.js]: indeed. I've always wondered why we need to pass in the client here as it's passed on from the `ApolloProvider` anyways, good to delete this here.

File: src/components/experiment-tracking/run-details-modal/run-details-modal.test.js
Comment (studioswong) [inline src/components/experiment-tracking/run-details-modal/run-details-modal.test.js:21]: Now that we have set up the mock for this mutation function, let's add the test that actually calls the `updateRunDetails` function ( which is called from `onApplyChanges` ), as well as mocking and testing the error state for the error message (i.e testing the existence of the `run-details-modal-error-wrapper` class) - otherwise there isn't any point of setting up this mock function here.

File: RELEASE.md
Comment (studioswong) [inline RELEASE.md:13]:
```suggestion
- Implemented the 'toggle bookmark' journey that allows bookmarking runs and displaying runs as a separate list . (#689)
```

File: src/components/experiment-wrapper/experiment-wrapper.js
Comment (studioswong) [inline src/components/experiment-wrapper/experiment-wrapper.js:86]: it would be more helpful to explain about what other components are here - I am guessing the run-metadata and run-details component?

**Review by studioswong — APPROVED** [review]
LGTM, thanks for getting through with the fix to the tests! 

I've left a few comments which I'll let you fix them before merging it in. For the tests for `run-details-modal`, I'll let you decide on whether to tackle it in this PR or as a separate PR.

**Pull request merged**
